v2/devdoc/logging.md

```diff
@@ -0,0 +1,339 @@
+// stack allocated contexts do not require any destructor
+#define LOG_CONTEXT_LOCAL_DEFINE(log_context, parent_context, ...) \
+    ...
+
+// when a user wants to specify the log context type name, the user can use this macro
+// otherwise, the default is to use the context variable name as type name
+// LOG_CONTEXT_TYPE_NAME(log_context_type_name)
+
+// when a user wants to specify properties for a context, they can use LOG_CONTEXT_PROPERTY
+// LOG_CONTEXT_PROPERTY(property_type, property_name, property_value)
+
+// when a user wants to specify string fields in a context using printf-style formatting, they can use LOG_CONTEXT_STRING_PROPERTY
+// LOG_CONTEXT_STRING_PROPERTY(property_name, property_conversion_specifier, property_value)
+```
+
+Various example usages:
+
+```c
+void log_from_a_function(LOG_CONTEXT_HANDLE log_context)
+{
+    LOGGER_LOG(LOG_LEVEL_ERROR, log_context, "log from a function!");
+}
+
+void examples(void)
+{
+    char* prop_value = malloc(2);
+    (void)strcpy(prop_value, "a");
+
+    // allocate a context dynamically
+    LOG_CONTEXT_HANDLE log_context;
+    LOG_CONTEXT_CREATE(log_context, NULL, LOG_CONTEXT_STRING_PROPERTY(property_name, "%s", MU_P_OR_NULL(prop_value)));
+
+    LOGGER_LOG(LOG_LEVEL_ERROR, log_context, "some_error with context");
+
+    LOG_CONTEXT_DESTROY(log_context);
+
+    // free the string, showing that a copy is made in the context
+    free(prop_value);
+
+    // chained allocated context
+    LOG_CONTEXT_HANDLE context_1;
+    LOG_CONTEXT_CREATE(context_1, NULL, LOG_CONTEXT_NAME(context_1), LOG_CONTEXT_STRING_PROPERTY(name, "%s", "haga"));
+    LOG_CONTEXT_HANDLE context_2;
+    LOG_CONTEXT_CREATE(context_2, context_1, LOG_CONTEXT_NAME(context_2), LOG_CONTEXT_STRING_PROPERTY(last name, "%s%s", "ua", "ga"), LOG_CONTEXT_PROPERTY(int32_t, age, 42));
+    
+    LOGGER_LOG(LOG_LEVEL_CRITICAL, context_1, "log with context 1, value of some_var = %d", 42);
+    LOGGER_LOG(LOG_LEVEL_VERBOSE, context_2, "log with context 2, some other string value is %s", "mumu");
+    
+    LOG_CONTEXT_DESTROY(context_1);
+
+    log_from_a_function(context_2);
+
+    // stack allocated context
+    LOG_CONTEXT_LOCAL_DEFINE(local_context, context_2, LOG_CONTEXT_STRING_PROPERTY(the_knights_that_say, "%s", "Nee!"));
+
+    LOGGER_LOG(LOG_LEVEL_VERBOSE, &local_context, "log with local context");
+    
+    // local context chaining another local context
+    LOG_CONTEXT_LOCAL_DEFINE(local_context_2, &local_context, LOG_CONTEXT_STRING_PROPERTY(other_knights_that_say, "%s", "Moo!"));
+    
+    LOGGER_LOG(LOG_LEVEL_VERBOSE, &local_context_2, "log with nee and moo");
+
+    log_from_a_function(&local_context_2);
+    
+    LOG_CONTEXT_DESTROY(context_2);
+}
+```
+
+The above example produces:
+
+```
+LOG_LEVEL_CRITICAL Time:Mon Aug 15 11:16:09 2022 File:G:\w\c-logging\v2\tests\logger_int\main.c:89 Func:main unnamed={ property_name=a  }  some critical error with context
```

Comment:
>LOG_LEVEL_CRITICAL [](http://example.com/codeflow?start=0&length=18)

why does line 61 (which I assume is the source of line 100) produce a LOG_LEVEL_CRITICAL when the line wants to log a LOG_LEVEL _ERROR?

Reply:
Synced up the example and output